src/pybullet_helpers/robots/single_arm.py

```diff
@@ -115,29 +187,76 @@ def closed_fingers(self) -> float:
 
-    @abc.abstractmethod
+    @cached_property
+    def initial_joints_state(self) -> JointsState:
+        """The joint values for the robot in its home pose."""
+        initial_joints_state = self.inverse_kinematics(self._ee_home_pose,
+                                                       validate=True)
+        # The initial joint values for the fingers should be open. IK may
+        # return anything for them.
+        initial_joints_state[self.left_finger_joint_idx] = self.open_fingers
+        initial_joints_state[self.right_finger_joint_idx] = self.open_fingers
+        return initial_joints_state
+
     def reset_state(self, robot_state: Array) -> None:
         """Reset the robot state to match the input state.
 
         The robot_state corresponds to the State vector for the robot
         object.
         """
-        raise NotImplementedError("Override me!")
+        rx, ry, rz, rf = robot_state
+        p.resetBasePositionAndOrientation(
+            self.robot_id,
+            self._base_pose.position,
+            self._base_pose.quat_xyzw,
+            physicsClientId=self.physics_client_id,
+        )
+        # First, reset the joint values to initial joint state,
+        # so that IK is consistent (less sensitive to initialization).
+        self.set_joints(self.initial_joints_state)
+
+        # Now run IK to get to the actual starting rx, ry, rz. We use
+        # validate=True to ensure that this initialization works.
+        joint_values = self.inverse_kinematics((rx, ry, rz), validate=True)
+        self.set_joints(joint_values)
+
+        # Handle setting the robot finger joints.
+        for finger_id in [self.left_finger_id, self.right_finger_id]:
+            p.resetJointState(self.robot_id,
+                              finger_id,
+                              rf,
+                              physicsClientId=self.physics_client_id)
 
-    @abc.abstractmethod
     def get_state(self) -> Array:
         """Get the robot state vector based on the current PyBullet state.
 
         This corresponds to the State vector for the robot object.
         """
-        raise NotImplementedError("Override me!")
+        ee_link_state = p.getLinkState(self.robot_id,
+                                       self.end_effector_id,
+                                       physicsClientId=self.physics_client_id)
+        rx, ry, rz = ee_link_state[4]
+        # Note: we assume both left and right gripper have the same joint
+        # position
+        rf = p.getJointState(
+            self.robot_id,
+            self.left_finger_id,
+            physicsClientId=self.physics_client_id,
+        )[0]
+        # pose_x, pose_y, pose_z, fingers
+        return np.array([rx, ry, rz, rf], dtype=np.float32)
 
-    @abc.abstractmethod
     def get_joints(self) -> JointsState:
         """Get the joints state from the current PyBullet state."""
-        raise NotImplementedError("Override me!")
+        joints_state = [
```

Comment:
use the new utilities here?

Reply:
We don't have utility for get joint states yet. Will come in future commits